### DISCUSSION TITLE
Support for an Every Operator (like $elemMatch)

### DISCUSSION THREAD

**Discussion by foopis23** [payloadcms/payload#1751] [Feature Requests & Ideas]
## Problem
Maybe I just can't think of a better way to write this query, but what I'm trying to do is make sure that every element in array is in another array. The best thing I came up with was get the inverse set and use the not_in operator.

## Suggested Feature
It would be really nice if we had an every operator. It could be used on array values to make sure all elements match a constraint.  I think it would be something like this.

```
payload.find({
    collection: "news",
    where: {
        tags: {
            every: {
                in: ["IT Services", "Engineering", "Support"]
            }
        }
    }
})
```

I think this should be an almost direct mapping to $elemMatch in mongodb, but I have never used $elemMatch and I haven't dug into the source code deep enough to see if that is even possible.

**Comment by foopis23** [comment]
An update on this. I didn't realize that $elemMatch matched any element. The operation I'm looking for is more like this in MongoDB.

```
"tags" {
    "$not": {
        "$elemMatch": {
            "$nin": ["IT Services", "Engineering", "Support"]
        }
    }
}
```

## Query Interface Options
Originally I was thinking "every" with a subquery makes sense. But after looking into it more, I think something like `every_in`, `every_equals`, `every_not_in`, and `every_not_equals` makes more sense. This has the benefit of being easier to understand and it fits a little better with how the query builder is actually working.

```
payload.find({
    collection: "news",
    where: {
        tags: {
            every_in: ["IT Services", "Engineering", "Support"]
        }
    }
})
```

## Implementation Details

### Query Builder
I already have a fork of payload where I have the original interface idea working for the multi-select field. It's a little hacked-in right now.  After taking an initial pass at this, what makes sense to me to make this more manageable is to add support for arrays in the operator map. So if we said
```
{
    'every_in': ['$not', '$elemMatch', '$nin']
}
```

Then it knows to make this object
```
{
    path: path,
    value: {
        $not: {
            $elemMatch: {
                $nin: val
            }
        }
    }
}
```

### GraphQL
Honestly, I don't have a lot of experience with Graphql, but I figure implementing something like this is just extending the graphql schema to support another operator.

**Reply by jmikrut** [reply]
Hey @foopis23 — I am onboard with your suggestions here and I'd say that you could probably take a crack at a PR for this new operator if you're up for it. The only thing I'd say is that it needs to work in all APIs, not just GraphQL, so you should make sure that it works with REST and Local APIs as well.

Will you give it a shot? Otherwise I'm sure we can get to it at some point but I am not sure when. Probably gonna be a bit as we have a lot on our plate for Q1!

**Reply by foopis23** [reply]
Yes, I can give it a shot. And sorry, I wasn't clear about that. I added Graphql at the end there to say, I don't really know how that part is going to work yet. I don't have a ton of experience with Graphql. I'm currently testing it with the Local and Rest API.

I will make sure it works with local, rest, grahpql on collections and globals for all of the field types that could have an array value before I PR.

**Comment by foopis23** [comment]
Made a PR for a first attempt at this
https://github.com/payloadcms/payload/pull/1776
